test/EventManagerIntrospectionTrait.php

```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace ZendTest\Mvc;
+
+use ReflectionProperty;
+use Zend\EventManager\EventManager;
+
+/**
+ * Offer methods for introspecting event manager events and listeners.
+ */
+trait EventManagerIntrospectionTrait
+{
+    public function getEventsFromEventManager(EventManager $events)
+    {
+        $r = new ReflectionProperty($events, 'events');
+        $r->setAccessible(true);
+        $listeners = $r->getValue($events);
+        return array_keys($listeners);
+    }
+
+    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
+    {
+        $r = new ReflectionProperty($events, 'events');
+        $r->setAccessible(true);
+        $listeners = $r->getValue($events);
+
+        if (! isset($listeners[$event])) {
+            return $this->traverseListeners([]);
+        }
+
+        return $this->traverseListeners($listeners[$event], $withPriority);
+    }
+
+    public function traverseListeners(array $queue, $withPriority = false)
```

Comment:
It's a generator. Suggest a better name?
On Oct 1, 2015 6:44 AM, "Marco Pivetta" <notifications@github.com> wrote:

> In test/EventManagerIntrospectionTrait.php
> <https://github.com/zendframework/zend-mvc/pull/31#discussion_r40904163>:
>
> > +    }
> > +
> > +    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
> > +    {
> > +        $r = new ReflectionProperty($events, 'events');
> > +        $r->setAccessible(true);
> > +        $listeners = $r->getValue($events);
> > +
> > +        if (! isset($listeners[$event])) {
> > +            return $this->traverseListeners([]);
> > +        }
> > +
> > +        return $this->traverseListeners($listeners[$event], $withPriority);
> > +    }
> > +
> > +    public function traverseListeners(array $queue, $withPriority = false)
>
> Method purpose is unclear
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zendframework/zend-mvc/pull/31/files#r40904163>.
>